webauthn/helpers/structs.py

```diff
@@ -1,45 +1,101 @@
+import base64
 from enum import Enum
 from typing import List, Literal, Optional
+from typing_extensions import Annotated
+
+from pydantic import (
+    ConfigDict,
+    BaseModel,
+    EncodedBytes,
+    EncoderProtocol,
+    FieldValidationInfo,
+    field_validator,
+)
+from pydantic_core import PydanticCustomError
 
-from pydantic import BaseModel, validator
-from pydantic.fields import ModelField
-
-from .bytes_to_base64url import bytes_to_base64url
 from .cose import COSEAlgorithmIdentifier
-from .json_loads_base64url_to_bytes import json_loads_base64url_to_bytes
 from .snake_case_to_camel_case import snake_case_to_camel_case
 
 
+class URLSafeBase64Encoder(EncoderProtocol):
+    """Base64 encoder."""
+
+    @classmethod
+    def decode(cls, data: bytes) -> bytes:
+        """Decode the data from URL safe base64 encoded bytes to original bytes data.
+
+        Args:
+            data: The data to decode.
+
+        Returns:
+            The decoded data.
+        """
+
+        try:
+            return base64.urlsafe_b64decode(data + b"====")
+        except ValueError as e:
+            raise PydanticCustomError(
+                "base64_decode", "Base64 decoding error: '{error}'", {"error": str(e)}
+            )
+
+    @classmethod
+    def encode(cls, value: bytes) -> bytes:
+        """Encode the data from bytes to a URL safe base64 encoded bytes.
+
+        Args:
+            value: The data to encode.
+
+        Returns:
+            The encoded data.
+        """
+
+        return base64.urlsafe_b64encode(value).replace(b"=", b"")
+
+    @classmethod
+    def get_json_format(cls) -> Literal["base64"]:
+        """Get the JSON format for the encoded data.
+
+        Returns:
+            The JSON format for the encoded data.
+        """
+        return "base64"
+
+
+URLSafeBase64 = Annotated[bytes, EncodedBytes(encoder=URLSafeBase64Encoder)]
```

Comment:
I feel like there's a lot going on here in this one line, what is this doing exactly?

Reply:
This is defining a type for Pedantic to use, with encoding/decoding logic "annotated" on. This is the new way to do custom field encoding/decoding in Pydantic v2. It's no longer possible to specify custom json encoding logic for the entire model